usr/src/lib/libzfs/common/libzfs_import.c

```diff
@@ -1064,35 +1064,67 @@ zpool_open_func(void *arg)
 }
 
 /*
- * Given a file descriptor, clear (zero) the label information.
+ * Given a file descriptor, a starting label and a number of labels to clear,
+ * invalidate or clear (zero) the label information.
  */
 int
-zpool_clear_label(int fd)
+zpool_clear_n_labels(int fd, unsigned int start, unsigned int n,
+    boolean_t force, boolean_t wipe)
 {
 	struct stat64 statbuf;
-	int l;
-	vdev_label_t *label;
+	unsigned int l, end;
+	vdev_label_t label;
 	uint64_t size;
 
+	char *buf = label.vl_vdev_phys.vp_nvlist;
+	size_t buflen = sizeof (label.vl_vdev_phys.vp_nvlist);
+
 	if (fstat64(fd, &statbuf) == -1)
 		return (0);
 	size = P2ALIGN_TYPED(statbuf.st_size, sizeof (vdev_label_t), uint64_t);
 
-	if ((label = calloc(sizeof (vdev_label_t), 1)) == NULL)
+	end = start + n;
+	if (end > VDEV_LABELS)
 		return (-1);
 
-	for (l = 0; l < VDEV_LABELS; l++) {
-		if (pwrite64(fd, label, sizeof (vdev_label_t),
-		    label_offset(size, l)) != sizeof (vdev_label_t)) {
-			free(label);
-			return (-1);
+	for (l = start; l < end; l++) {
+		if (!(force && wipe)) {
+			if (pread64(fd, &label, sizeof (vdev_label_t),
+			    label_offset(size, l)) != sizeof (vdev_label_t))
+				return (-1);
+
+			if (!force) {
+				nvlist_t *config = NULL;
+				if (nvlist_unpack(buf, buflen, &config, 0) != 0)
+					return (-1);
+				nvlist_free(config);
+			}
+		}
+
+		if (wipe) {
+			(void) memset(&label, 0, sizeof (vdev_label_t));
+		} else {
+			if (nvlist_invalidate(buf, buflen) != 0)
```

Comment:
If the goal is to prevent zpool import from showing the pool, then we should set the TXG value to 0 which is how ZFS clears a label while still leaving some information around for debugging.

Reply:
The main goal of the patch is to allow wiping a single label with *minimal* changes.

Acting on nvs encoding type (introducing a new type "invalid") allows touching a single byte, which is great because it minimizes the risks of damaging another FS that would have been created over the label. Acting on txg would touch 8 bytes and greatly improve the chances to break something. Also, as a side effect, acting on nvs encoding type keeps last txg value available for debugging.